contracts/BonusRewards.sol

```diff
@@ -3,11 +3,340 @@
+  }
+
+  function getResponders() external view override returns (address[] memory) {
+    return responders;
+  }
+
+  /// @notice update pool's bonus per staked token till current block timestamp
+  function updatePool(address _lpToken) public override {
+    Pool storage pool = pools[_lpToken];
+
+    uint256 poolLastUpdatedAt = pool.lastUpdatedAt;
+    if (poolLastUpdatedAt == 0 || block.timestamp <= poolLastUpdatedAt) return;
+    uint256 lpTotal = IERC20(_lpToken).balanceOf(address(this));
+    if (lpTotal == 0) {
+      pool.lastUpdatedAt = block.timestamp;
+      return;
+    }
+
+    for (uint256 i = 0; i < pool.bonuses.length; i ++) {
+      Bonus storage bonus = pool.bonuses[i];
+      if (pool.lastUpdatedAt < bonus.endTime && bonus.startTime < block.timestamp) {
+        uint256 bonusForTime = _calRewardsForTime(bonus, pool.lastUpdatedAt);
+        bonus.accRewardsPerToken = bonus.accRewardsPerToken + bonusForTime / lpTotal;
+      }
+    }
+    pool.lastUpdatedAt = block.timestamp;
+  }
+
+  function claimRewards(address _lpToken) public override {
+    User memory user = users[_lpToken][msg.sender];
+    if (user.amount == 0) return;
+
+    updatePool(_lpToken);
+    _claimRewards(_lpToken, user);
+    _updateUserWriteoffs(_lpToken);
+  }
+
+  function claimRewardsForPools(address[] calldata _lpTokens) external override {
+    for (uint256 i = 0; i < _lpTokens.length; i++) {
+      claimRewards(_lpTokens[i]);
+    }
+  }
+
+  function deposit(address _lpToken, uint256 _amount) external override nonReentrant notPaused {
+    require(pools[_lpToken].lastUpdatedAt > 0, "Blacksmith: pool does not exists");
+    require(IERC20(_lpToken).balanceOf(msg.sender) >= _amount, "Blacksmith: insufficient balance");
+
+    updatePool(_lpToken);
+    User storage user = users[_lpToken][msg.sender];
+    _claimRewards(_lpToken, user);
+    user.amount = user.amount + _amount;
+    _updateUserWriteoffs(_lpToken);
+
+    IERC20(_lpToken).safeTransferFrom(msg.sender, address(this), _amount);
+    emit Deposit(msg.sender, _lpToken, _amount);
+  }
+
+  /// @notice withdraw up to all user deposited
+  function withdraw(address _lpToken, uint256 _amount) external override nonReentrant notPaused {
+    updatePool(_lpToken);
+
+    User storage user = users[_lpToken][msg.sender];
+    _claimRewards(_lpToken, user);
+    uint256 amount = user.amount > _amount ? _amount : user.amount;
+    user.amount = user.amount - amount;
+    _updateUserWriteoffs(_lpToken);
+
+    _safeTransfer(_lpToken, amount);
+    emit Withdraw(msg.sender, _lpToken, amount);
+  }
+
+  /// @notice withdraw all without rewards
+  function emergencyWithdraw(address _lpToken) external override nonReentrant {
+    User storage user = users[_lpToken][msg.sender];
+    uint256 amount = user.amount;
+    user.amount = 0;
+    _safeTransfer(_lpToken, amount);
+    emit Withdraw(msg.sender, _lpToken, amount);
+  }
+
+  /// @notice called by authorizers only
+  function addBonus(
```

Comment:
There is no way to amend weeklyRewards or endTime once the bonus is added. This feels like it would be easy to make a mistake (maybe wrong number of decimals) and have huge/tiny rewards that can't be changed.

Reply:
added `updateBonus` to allow updating the `startTime` and `weeklyRewards` under certain conditions.

Reply:
Update bonus checks if bonus.endTime > block.timestamp before allowing an update. This means you still cannot update a bonus that is currently running. I will comment on it directly

Reply:
Update bonus looks good. Here are two things to keep in mind:

1 - you cannot decrease rewards. So if you add the wrong amount in addBonus or extendBonus it cannot be reduced. An example when this might happen is by getting decimals wrong.

2 - If there is somehow two bonuses of the same type active at the same time it will update both. I don't think this is possible but worth thinking about. We can have two bonuses of the same address added to the lpToken but I do not think they can both be active at the same time.

Reply:
1. Wdym it cannot decrease? It should be able to based on the following calculation. The endTime will be updated based on the new WeeklyRewards.
`bonus.endTime = uint48(remBonusToDistribute * WEEK / _weeklyRewards + block.timestamp);`
We are able to update the weekly Rewards on Kovan back and forth with the endtime updated as expected.
https://testcoverprotocol.web.app/app/rewards

2. Yeah, it is impossible to have two bonus program on the same lpToken pool and same bonus token at the same time.

Reply:
> 1. It should be able to decrease based on the following calculation. The endTime will be updated based on the new WeeklyRewards.

What I mean is that you cannot decrease total rewards (remBonusToDistribute). If you add 100x too much reward token by accident you cannot retrieve it. I don't think that is a big issue but is something to be careful when adding bonuses. You are right that you can reduce weekly emissions.

Reply:
Oh yeah, once transferred in, cannot be out. Good point tho. It should be a very rare case. The farmers will benefit.